### PR TITLE
Context api alternative implementation 

### PR DESCRIPTION
**Pull request by satac2**
Links the context objects together as a linked list where each time new data is written to a context, a new node is added containing that new information and each node has access to all the data before it and none of the data after. Within each node there is another linked list actually containing the key value pairs. This implementation uses no STL and therefore I believe should have no ABI compatibility issues. @reyang @maxgolov.

### CODE REVIEW

File: api/test/context/context_test.cc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  M m2 = {{"other_key", "789"}};
+
+  context::Context* foo_context   = test_context.SetValues(m1);
+  context::Context* other_context = foo_context->SetValues(m2);
+
+  EXPECT_EQ(nostd::get<nostd::string_view>(other_context->GetValue("test_key")), "123");
+  EXPECT_EQ(nostd::get<nostd::string_view>(other_context->GetValue("foo_key")), "456");
+}
+
+// Tests that copying a context copies the key value pairs as expected.
```
Comment (reyang) [inline api/test/context/context_test.cc:87]: Might be good to have a test with duplicated keys and verify the last winner.

File: api/include/opentelemetry/context/context.h [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+      Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+      {
+        return new Context(key, value, this);
+      }
+
+      // Returns the value associated with the passed in key.
+      common::AttributeValue GetValue(nostd::string_view key)
+      {
+        //Iterate through the context nodes
+        for(Context* context = this; context != nullptr; context = context->next_){
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:60]: There might be potential optimization where the `DataList` nodes are shared and ref-counted. Not blocking this PR though.

File: api/test/context/context_test.cc (hunk 1 above)
Comment (satac2) [reply api/test/context/context_test.cc:87]: Isn't that case already covered with the ContextKeyOverwrite test?

Comment (reyang) [reply api/test/context/context_test.cc:87]: Yep, you're right, somehow I missed it.

File: api/include/opentelemetry/context/context.h [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+              return data->value_;
+            }
+          }
+        }
+        return "";
+      }
+
+      // Iterates through the current and comparing context objects
+      // to check for equality.
+      bool operator==(const Context &other)
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:73]: Do we need this comparison operator?

File: api/include/opentelemetry/context/context.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        if(next_ != other.next_){
+          return false; 
+        }
+
+        DataList* other_data = other.head_;
+
+        // Iterate through the lists and compare the keys
+        // TODO should also compare the values but the nostd::variant
+        // seems to not have a comparison operator
+        for(DataList* data = head_; data != nullptr; data = data->next_){
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:86]: Let's rather implement a dedicated comparator on `DataList`. Also, besides not taking values into account, the code here assumes that the keys are ordered in the same way.

Comment (satac2) [reply api/include/opentelemetry/context/context.h:86]: When you are comparing two contexts what exactly are you checking for? I was expecting that two contexts with differently ordered keys would be considered different contexts. Also I have been having problems with nostd::variant comparison, I am planning on looking into it after getting the context dealt with.

File: api/include/opentelemetry/context/context.h (hunk 3 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:73]: I think so, I am planning on using it within the threadlocal_context wrapper to ensure that the user is detaching the context that they are expecting.

File: api/include/opentelemetry/context/context.h [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+      // Accepts a new iterable and then returns a new context that
+      // contains the new key and value data.
+      Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+      {
+        return new Context(key, value, this);
+      }
+
+      // Returns the value associated with the passed in key.
+      common::AttributeValue GetValue(nostd::string_view key)
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Yes, that change has been made.

File: api/include/opentelemetry/context/context_value.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+  using ContextValue = nostd::variant<bool,
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/context/context_value.h]: None to come to my mind, @reyang do you have any thoughts?

[the remark this answers is not in the document]
Comment (reyang) [reply api/include/opentelemetry/context/context_value.h]: Later we might need to add `Baggage` but the list provided by @pyohannes should be good enough for now.

File: api/include/opentelemetry/context/context.h [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+public:
+  nostd::shared_ptr<Context> *context_ptr_;
+
+  // Creates a new context by calling a constructor with the passed in
+  // keyValueIterable and returns a shared_ptr pointing to it.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  static nostd::shared_ptr<Context> CreateContext(T &keys_and_values)
+  {
+    nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(keys_and_values)};
+    temp->context_ptr_              = &temp;
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:31]: I don't fully understand the use of `context_ptr_`. It seems it is a pointer to a shared pointer of the object itself. How is it different from just accessing `this`?

Comment (satac2) [reply api/include/opentelemetry/context/context.h:31]: Because I want both the user and the node right behind this node to be using the same shared_ptr.

Comment (satac2) [reply api/include/opentelemetry/context/context.h:31]: If I just accessed this when creating a shared_ptr to return to the user and a shared_ptr to point to the next node, then wouldn't they not be linked together and attempt to deconstruct the context twice?

Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:31]: Consider using [enable_shared_from_this](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) for this use case. A pointer to a shared pointer looks dangerous.

Comment (satac2) [reply api/include/opentelemetry/context/context.h:31]: Perhaps it is what is cause my window test failures.

Comment (satac2) [reply api/include/opentelemetry/context/context.h:31]: Thanks, i'll give it a try, I assume it works smoothly with nostd::shared_ptr?

Comment (satac2) [reply api/include/opentelemetry/context/context.h:31]: But in that case wouldn't I need to keep a shared_ptr pointing to the object as a member of the object? Doesn't that defeat the purpose?

Comment (satac2) [reply api/include/opentelemetry/context/context.h:31]: I am on board, will be rewriting to avoid this issue.

File: api/include/opentelemetry/context/context.h [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    return temp;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  nostd::shared_ptr<Context> SetValues(T &values) noexcept
+  {
+    nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(values)};
+    temp->context_ptr_              = &temp;
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:50]: I think storing `context_ptr_` like this is dangerous: it points to a `nostd::shared_ptr` on the stack, this might get copied around and the address it points to can be invalid at a later point in time.

You could get around the need of storing `context_ptr_` by making `SetValues` methods static, or by taking them out of the class as functions. Then you can pass in a parent as a `nostd::shared_ptr` explicitly:
```c
template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
nostd::shared_ptr<Context> SetValues(nostd::shared_ptr<Context> parent, T &values) noexcept
{
  nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(values)};
  temp->context_ptr_              = &temp;
  temp->SetNext(parent);
  return temp;
}
```

File: api/include/opentelemetry/context/context.h (hunk 7 above)
Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:31]: You can disregard my comment, I think using `enable_shared_from_this` is ABI stable.

I tried to come up with another suggestion [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/180/files#r457619122).

File: api/include/opentelemetry/context/context.h (hunk 8 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: I modified the implementation considerably it now avoids the issue, could I convince you to take a look at the code again?

File: api/test/context/context_test.cc [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+  EXPECT_EQ(nostd::get<nostd::string_view>(foo_context.GetValue(string_view_test)), "123");
+}
+
+// Tests that the original context does not change when a value is
+// written to it.
+TEST(ContextTest, ContextImmutability)
+{
+  std::map<std::string, context::ContextValue> map_test = {{"test_key", "123"}};
+  std::map<std::string, context::ContextValue> foo_map  = {{"foo_key", "456"}};
```
Comment (pyohannes) [inline api/test/context/context_test.cc:56]: I think `foo_map` is not used.

File: api/include/opentelemetry/context/context.h [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+      }
+    }
+
+    // Builds a data list with just a key and value, so it will just be the head
+    // and returns that head.
+    DataList(nostd::string_view key, ContextValue value)
+    {
+      key_        = new char[nostd::string_view(key).size()];
+      key_length_ = nostd::string_view(key).size();
+      memcpy(key_, nostd::string_view(key).data(), nostd::string_view(key).size() * sizeof(char));
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:133]: No need for all the explicit `nostd::string_view` constructors here. You can just use `key.size()` and `key.data()`.

File: api/include/opentelemetry/context/context.h [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+      value_ = iter->second;
+      next_  = nostd::shared_ptr<DataList>{nullptr};
+      ++iter;
+
+      DataList *previous_node = this;
+      // Iterate over the keys and values iterable and add nodes
+      for (; iter != std::end(keys_and_vals); ++iter)
+      {
+        nostd::shared_ptr<DataList> node = nostd::shared_ptr<DataList>{new DataList()};
+        node->next_                      = nostd::shared_ptr<DataList>{nullptr};
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:116]: You can leave out this line, as you set this in the `DataList` constructor.

File: api/include/opentelemetry/context/context.h [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        node->key_                       = new char[nostd::string_view(iter->first).size()];
+        node->key_length_                = nostd::string_view(iter->first).size();
+        memcpy(node->key_, nostd::string_view(iter->first).data(),
+               nostd::string_view(iter->first).size() * sizeof(char));
+        node->value_         = iter->second;
+        previous_node->next_ = node;
+        previous_node        = &*node;
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:123]: You can make this a `previous_node = node`.

File: api/include/opentelemetry/context/context.h [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    size_t key_length_;
+
+    ContextValue value_;
+
+    DataList() { next_ = nullptr; }
+
+    // Builds a data list off of a key and value iterable and returns the head
+    template <class T,
+              nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+    DataList(const T &keys_and_vals)
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:99]: Please add a test for initializing this with an empty map (for `keys_and_vals`). I think this will fail.

File: api/include/opentelemetry/context/context.h [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    {
+      if (key.size() == data->key_length_)
+      {
+        if (memcmp(key.data(), data->key_, data->key_length_) == 0)
+        {
+          return data->value_;
+        }
+      }
+    }
+    return "";
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:76]: We need a way to check whether a key exists. Checking for an empty string is not sufficient, as empty strings are valid values.

File: api/include/opentelemetry/context/context_value.h [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+using ContextValue = nostd::variant<bool,
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context_value.h:13]: Please also consider [this comment](https://github.com/open-telemetry/opentelemetry-cpp/pull/180#discussion_r456120855).

File: api/include/opentelemetry/context/context.h [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+// the list that determines which keys and values it has access to. All that
+// come before and none that come after.
+class Context
+{
+
+public:
+  // Creates a context object from a map of keys and identifiers, this will
+  // hold a shared_ptr to the head of the DataList linked list
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &keys_and_values)
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:25]: Will we need the functionality to retrieve a parent context for a context? Will this be needed for modifying the current context at some point?

File: api/include/opentelemetry/context/context.h (hunk 14 above)
Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:76]: Maybe adding a `HasKey` method?

File: api/include/opentelemetry/context/context.h (hunk 12 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:123]: My goal he was to avoid making previous_node a shared_ptr, because I don't have access to the other shared_ptr pointing to this object at the time, and creating another would cause the node to destruct too early.

File: api/include/opentelemetry/context/context.h (hunk 16 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:25]: I can't think of an instance, can you?

File: api/include/opentelemetry/context/context.h (hunk 12 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:123]: *here

File: api/include/opentelemetry/context/context_value.h (hunk 15 above)
Comment (satac2) [reply api/include/opentelemetry/context/context_value.h:13]: When I try to add nostd::unique_ptr<char []>.
I get 'opentelemetry::v0::nostd::variant<Ts>::variant(const opentelemetry::v0::nostd::variant<Ts>&) [with Ts = {bool, long int, 
long unsigned int, double, opentelemetry::v0::nostd::unique_ptr<char []>}]' is implicitly deleted because the default definition would be ill-formed:
   variant(const variant &) = default;

Comment (satac2) [reply api/include/opentelemetry/context/context_value.h:13]: and the same for nostd::shared_ptr<SpanContext>. Do you have any idea why that is?

File: api/test/context/context_test.cc (hunk 9 above)
Comment (satac2) [reply api/test/context/context_test.cc]: removed.

File: api/include/opentelemetry/context/context.h (hunk 10 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Removed.

File: api/include/opentelemetry/context/context.h (hunk 11 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Removed.

File: api/include/opentelemetry/context/context.h (hunk 13 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:114]: Added test and fixed functionality.

File: api/include/opentelemetry/context/context.h (hunk 14 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Added HasKey method and test.

File: api/include/opentelemetry/context/context_value.h [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+using ContextValue = nostd::variant<bool, int64_t, uint64_t, double>;
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context_value.h:13]: I think we'll need `nostd::shared_ptr<SpanContext>` too.

File: api/include/opentelemetry/context/context.h (hunk 16 above)
Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:25]: I'm not 100% sure, maybe I'm missing some part of the high-level view.

I think we'll use `Context` to obtain the currently active span: we will have a method to obtain the current context, the current context will hold the `SpanContext` of the active span. I'm trying to figure out necessary steps to make this possible:

1. When a new span starts, we need to set a new current context. We will be able to get this new context via `SetValue("current_span", span_context)`. Then the new span will be stored in the new current context.
2. When the span ends, we will need to restore the previous context, as this should be the current context now (returning the parent span as active span).

I think we'll need some kind of parenting of contexts, but I'm not sure if it will be managed by the `Context` itself or by some other class.

File: api/include/opentelemetry/context/context_value.h (hunk 17 above)
Comment (satac2) [reply api/include/opentelemetry/context/context_value.h:13]: When I tried to add it I got a: variant is implicitly deleted because the default definition would be ill-formed error

File: api/include/opentelemetry/context/context.h (hunk 16 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:25]: I believe that will be handled within the threadlocal context class

Comment (satac2) [reply api/include/opentelemetry/context/context.h:25]: The implementation of which is described by the context spec

Comment (satac2) [reply api/include/opentelemetry/context/context.h:25]: Or rather, interface of

File: api/include/opentelemetry/context/context_value.h (hunk 17 above)
Comment (satac2) [reply api/include/opentelemetry/context/context_value.h:13]: So the nostd::shared_ptr<SpanContext> does work. But whenever I try to add a unique_ptr of any type, I am getting the implicitly deleted because of ill formed definition error.

File: api/include/opentelemetry/context/context.h [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data. It attaches the
+  // exisiting list to the end of the new list.
+  template <class T>
+  Context SetValues(T &values) noexcept
+  {
+    Context context = Context(values);
+    nostd::shared_ptr<DataList> last_node;
+    for (nostd::shared_ptr<DataList> data = context.head_; data != nullptr; data = data->next_)
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:45]: Do we need `shared_ptr` while iterating through the linked list or there can be an optimization?

File: api/include/opentelemetry/context/context.h [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+    {
+      if (key.size() == data->key_length_)
+      {
+        if (memcmp(key.data(), data->key_, data->key_length_) == 0)
+        {
+          return data->value_;
+        }
+      }
+    }
+    return (int64_t)0;
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:76]: Not blocking this PR, do we need a nil type?

File: api/include/opentelemetry/context/context.h (hunk 18 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h:45]: Changed to just a ptr.

File: api/include/opentelemetry/context/context_value.h (hunk 17 above)
Comment (satac2) [reply api/include/opentelemetry/context/context_value.h]: nostd::shared_pr<SpanContext> has been added is the unique_ptr necessary?

File: api/include/opentelemetry/context/context.h [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+      {
+        key_ = nullptr;
+      }
+    }
+
+    // Builds a data list with just a key and value, so it will just be the head
+    // and returns that head.
+    DataList(nostd::string_view key, ContextValue value)
+    {
+      key_        = new char[nostd::string_view(key).size()];
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:152]: No need for the explicit `nostd::string_view` here, this can just be `key.size()`.

File: api/include/opentelemetry/context/context.h (hunk 13 above)
Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:114]: Please add an additional move constructor:
```c++
DataList &operator=(DataList &&other)
{
  key_length_ = other.key_length_;
  value_      = std::move(other.value_);
  next_       = std::move(other.next_);

  key_       = other.key_;
  other.key_ = nullptr;

  return *this;
}
```

Then you can simplify this constructor and avoid code duplication:
```c++
template <class T>
DataList(const T &keys_and_vals) : key_{nullptr}
{
  bool first = true;
  auto *node = this;
  for (auto &iter : keys_and_vals)
  {
    if (first)
    {
      *node = std::move(DataList(iter.first, iter.second));
      first = false;
    } 
    else
    {
      node->next_ = nostd::shared_ptr<DataList>(new DataList(iter.first, iter.second));
      node        = node->next_.get();
    } 
  } 
}

File: api/include/opentelemetry/context/context.h (hunk 19 above)
Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:76]: Boost has a [boost::blank](https://www.boost.org/doc/libs/1_68_0/boost/blank.hpp) type for this purpose. The more C++ idiomatic way would be top wrap the return value in a [std::optional](https://en.cppreference.com/w/cpp/utility/optional), but if feel skeptical about extending `nostd` by this (it only was introduced in C++17).

I also think this shouldn't block this PR.

File: api/include/opentelemetry/context/context.h [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+  // exisiting list to the end of the new list.
+  Context SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context context      = Context(key, value);
+    context.head_->next_ = head_;
+    return context;
+  }
+
+  // Returns the value associated with the passed in key.
+  context::ContextValue GetValue(nostd::string_view key)
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:64]: This can be `const` and `noexcept`.

File: api/include/opentelemetry/context/context.h [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+        {
+          return data->value_;
+        }
+      }
+    }
+    return (int64_t)0;
+  }
+
+  // Checks for key and returns true if found
+  bool HasKey(nostd::string_view key)
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:80]: This can be `const` and `noexcept`.

**Comment by reyang** [comment]
The change looks good enough to me.

I think we might need some memory testing (e.g. checking if all the allocated linked-list nodes are freed up before exiting the application (e.g. by using our custom malloc/free) to make sure we're actually doing the correct/efficient ref-counting. Probably in another PR.

Looks like there are still open questions/comments, I'll wait until we got approval from @pyohannes.

File: api/include/opentelemetry/context/context.h [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+  {
+    Context context      = Context(key, value);
+    context.head_->next_ = head_;
+    return context;
+  }
+
+  // Returns the value associated with the passed in key.
+  context::ContextValue GetValue(nostd::string_view key)
+  {
+    for (DataList *data = &*head_; data != nullptr; data = &*(data->next_))
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:66]: I think this is a bit more readable:
```c++
for (DataList *data = head_.get(); data != nullptr; data = data->next_.get())
```

File: api/include/opentelemetry/context/context.h (hunk 18 above)
Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:45]: This can be simplified a bit:
```c++
nostd::shared_ptr<DataList>& last = context.head_;                          
while (last->next_ != nullptr) { last = last->next_; } 
                                                                                 
last->next_ = head_; 
```

File: api/include/opentelemetry/context/context.h [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+        }
+      }
+    }
+    return (int64_t)0;
+  }
+
+  // Checks for key and returns true if found
+  bool HasKey(nostd::string_view key)
+  {
+    for (nostd::shared_ptr<DataList> data = head_; data != nullptr; data = data->next_)
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:82]: We can avoid copies of `shared_ptr` here:
```c++
for (DataList *data = head_.get(); data != nullptr; data = data->next_.get())
```

**Comment by pyohannes** [comment]
> I think we might need some memory testing (e.g. checking if all the allocated linked-list nodes are freed up before exiting the application (e.g. by using our custom malloc/free) to make sure we're actually doing the correct/efficient ref-counting. Probably in another PR.

I ran the unit test through valgrind, and it looks fine. It would be nice to have some automated memory checking though in our CI.

File: api/include/opentelemetry/context/context.h [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+    DataList(const T &keys_and_vals)
+    {
+      if (std::begin(keys_and_vals) != std::end(keys_and_vals))
+      {
+        auto iter = std::begin(keys_and_vals);
+        // Create list head
+        key_        = new char[nostd::string_view(iter->first).size()];
+        key_length_ = nostd::string_view(iter->first).size();
+        memcpy(key_, nostd::string_view(iter->first).data(),
+               nostd::string_view(iter->first).size() * sizeof(char));
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Removed.

File: api/include/opentelemetry/context/context.h (hunk 20 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Removed.

File: api/include/opentelemetry/context/context.h (hunk 21 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Added.

File: api/include/opentelemetry/context/context.h (hunk 22 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Added.

File: api/include/opentelemetry/context/context.h (hunk 23 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Change made.

File: api/include/opentelemetry/context/context.h (hunk 24 above)
Comment (satac2) [reply api/include/opentelemetry/context/context.h]: Change made.

**Comment by satac2** [comment]
Isn't there memory checking in the CI?

File: api/include/opentelemetry/context/context.h [hunk 26]
```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
```
Comment (pyohannes) [inline api/include/opentelemetry/context/context.h:8]: Some include nits: I think `atomic` and `key_value_iterable_view.h` are not needed.

Comment (satac2) [reply api/include/opentelemetry/context/context.h]: I have removed them.

**Comment by pyohannes** [comment]
Please rebase and then @reyang please merge.
